### PR TITLE
API to construct `InputType`s for Record types.

### PR DESCRIPTION
**Pull request by eamsden**
Adds the `RecordInputType` type and associated functions `inputFieldWith`, `inputField`, and `inputRecord`.

These `inputField{,With}` functions and the `Contravariant` and `Divisible` operations create `RecordInputType` values which describe how a Haskell record type may be encoded into a Dhall expression. The `inputRecord` function converts a `RecordInputType` to an `InputType`.

An example (using the same type as the example for `RecordType`) is given in the documentation comment for `RecordInputType.

See #525

### CODE REVIEW

File: src/Dhall.hs [hunk 1]
```diff
@@ -1333,3 +1340,80 @@ field key valueType =
...
+> }
+
+    Our injector has type 'InputType' @Project@, but we can't build that out of any
+    smaller injectors, as 'InputType's cannot be combined (they are only 'Contravariant's).
+    However, we can use an 'InputRecordType' to build an 'InputType' for @Project@:
+
+> injectProject :: InputType Project
+> injectProject =
+>   inputRecord
+>     (  adapt >$< inputField "name" (inject @Text)
```
Comment (Gabriel439) [inline src/Dhall.hs]: I believe the type applications are not necessary since they should be inferred by the context

File: src/Dhall.hs [hunk 2]
```diff
@@ -1333,3 +1340,80 @@ field key valueType =
...
+>              >*< inputField "description" (inject @Text)
+>              >*< inputField "stars" (inject @Natural)
+>     )
+>   where
+>     adapt (Project{..}) = (projectName, (projectDescription, projectStars))
+
+-}
+
+-- | Infix 'contramap'
+(>$<) :: Contravariant f => (a -> b) -> f b -> f a
```
Comment (Gabriel439) [inline src/Dhall.hs]: I just realized that the `>$<` operator already exists in the `contravariant` package:

http://hackage.haskell.org/package/contravariant-1.5/docs/Data-Functor-Contravariant.html

... so we can reuse that

Comment (eamsden) [reply src/Dhall.hs]: I can drop that definitiion and just rexport the contravariant definition. Give me a few minutes.

**Comment by eamsden** [comment]
@quasicomputational maybe we should use the `inject` prefix? I was naming after `InputType`, which they are intended to aid construction of.

calling them `recordField{,With}` doesn't really give an idea of their distinction from `field`

File: src/Dhall.hs (hunk 1 above)
Comment (eamsden) [reply src/Dhall.hs]: I think you are right. And actually, given the way the API is written, they should either be `inputFieldWith` or the second parameter should be dropped entirely. The example as written woudn't typecheck!

Comment (Gabriel439) [reply src/Dhall.hs]: Note that we have `doctest` enabled on this module, so if you transform the example into a doctest then it will be checked 🙂

Comment (eamsden) [reply src/Dhall.hs]: I might do that in a separate PR, as it would involve modifying the $setup block to bring the `Project` type into scope.

Comment (Gabriel439) [reply src/Dhall.hs]: Yeah, it's not urgent to do in this PR

File: src/Dhall.hs [hunk 3]
```diff
@@ -1333,3 +1340,87 @@ field key valueType =
...
+>   where
+>     adapt (Project{..}) = (projectName, (projectDescription, projectStars))
+
+-}
+
+-- | Infix 'divided'
+(>*<) :: Divisible f => f a -> f b -> f (a, b)
+(>*<) = divided
+
+infixr 4 >*<
```
Comment (eamsden) [inline src/Dhall.hs]: I think this should be `infixl 4 >*<` to work properly with the `>$<` exported from contravariant, right?

Comment (Gabriel439) [reply src/Dhall.hs]: It should still be `infixr` if you want the tuple nesting to be `(a, (b, c))`, but the precedence needs to be elevated to 5

**Comment by eamsden** [comment]
@Gabriel439 does this need anything else before it can be merged?

**Comment by Gabriel439** [comment]
@eamsden: No, nothing needs to be done.  I just thought you had additional changes planned, but if you think this is ready then I'll go ahead and merge

**Pull request merged**